---
title: "Cohort_77_WGCNA_analysis_VizAnnotation"
author: "Nick Kron"
date: "9/27/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<!-- Load Packages -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}
library(ggplot2)
library(RColorBrewer)
library(cowplot)

library(WGCNA)

library(tidyr)
library(tibble)
library(dplyr)

library(ComplexHeatmap)
library(circlize)

```


<!-- Define function to exract tx names -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}

library("stringr")

extract_tx <- function(x){
  if(typeof(x) == "character"){
  stringr::str_extract(x, '[NX][MR]_[0-9]+.[0-9]')
  }
  else{
    print('ERROR: input is not a vector of type character')
  }
}

```

<!-- Define function to summarize count table to a group level -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}
summarize_group <- function(ma, group){
    counts_group = t(sapply(rownames(ma), function(g){
        sapply(levels(group), function(i){
            idx = which(group == i)
            mean(ma[g, idx], na.rm = TRUE)
        })
    }))
    colnames(counts_group) = levels(group)
    counts_group
}
```

<!-- function to generate Cytoscape files -->
```{r, eval= TRUE, echo=FALSE, message=FALSE, warning = FALSE}
exportAll2Cytoscape <- function(cols, TOM, data, minweight = -1, maxedges = 30000){
  modules = cols
  modules = unique(modules)
  modules[modules != "grey"]
  
  
  
  lapply(modules, function(x){
    probes = names(data)
    inModule = is.finite(match(moduleColors, x))
    modProbes = probes[inModule]    
    #modProbes = extract_tx(modProbes)
    modGenes = annot$human_gene[match(modProbes, annot$Ac_ref_Tx)]
    # Select the corresponding Topological Overlap
    modTOM = TOM[inModule, inModule]
    dimnames(modTOM) = list(modProbes, modProbes)
    nGenes = length(modProbes)
    edgeprob = (maxedges/(nGenes*nGenes))
    if(edgeprob >1){
      edgeprob = 1
    }
    if(edgeprob <0){
      edgeprob = 1
    }
    
    if(minweight <0 | minweight*(nGenes*nGenes) > maxedges){
      weight = quantile(modTOM,
                            probs = 1- edgeprob,
                            names = FALSE)
    }
    else{
      weight = quantile(modTOM,
                            probs = minweight,
                            names = FALSE)
    }
    print(c(edgeprob, weight, maxedges/edgeprob ))
    # Export the network into edge and node list files Cytoscape can read
    WGCNA::exportNetworkToCytoscape(modTOM,
                                    edgeFile = paste(paste0(projdir,"Cytoscape/Cohort_77_CytoscapeInput_edges_"), paste(x), ".txt", sep=""),
                                    nodeFile = paste(paste0(projdir,"Cytoscape/Cohort_77_CytoscapeInput_nodes_"), paste(x), ".txt", sep=""),
                                    weighted = TRUE,
                                    threshold = weight,
                                    nodeNames = modProbes,
                                    altNodeNames = modGenes,
                                    nodeAttr = moduleColors[inModule])
  })
  
}


```

<!-- function to generate DyNetViewer expression files -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}

genCytoscapeTPM <- function(abundances, group, cols){
  modules <- unique(cols)
  group <- unique(group)
  
  lapply(modules, function(x){
    edgeFile = paste(paste0(projdir,"Cytoscape/Cohort_77_CytoscapeInput_edges_"), paste(x), ".txt", sep="")
    nodeFile = paste(paste0(projdir,"Cytoscape/Cohort_77_CytoscapeInput_nodes_"), paste(x), ".txt", sep="")
    outFile = paste(paste0(projdir,"Cytoscape/Cohort_77_CytoscapeInput_exprDat_"), paste(x), ".txt", sep="")
    nodes <- read.table(file = nodeFile, sep = "\t", header = TRUE)
    edges <- read.table(file = edgeFile, sep = "\t", header = TRUE)
    
    dat <- abundances %>% as.data.frame() %>%
    rownames_to_column(var = "nodeName") %>% 
    inner_join(nodes %>% dplyr::select(nodeName)) %>%
    column_to_rownames("nodeName") %>%
    droplevels() %>%
    as.matrix()
    
    node_TPM <- summarize_group(ma = dat, group = group) %>% as.data.frame() %>% rownames_to_column(var = "nodeName")
    write.table(file = outFile, x = node_TPM, quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t" )
    
  })
  
  
}

```

<!-- Load data -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}

projdir = ("../results")

annot = read.delim(file = "../data/annotation_data/AcTx2HsUniProt.tab", header = TRUE, sep = "\t")
load(file = "../data/r_data/annot.R")

```


##Load sample data
```{r, echo=FALSE, message=FALSE, warning = FALSE}
lnames= load(file = "../data/r_data/Consensus_dataInput.R")
lnames= load(file = "../data/r_data/Consensus_NetworkConstruction_man.R")

samples <- read.csv(file = "../data/metadata/metadata.csv", header = TRUE)

exprDat <- t(rbind(multiExpr[[1]]$data, multiExpr[[2]]$data))

tx2module <- data.frame(tx = rownames(exprDat), module = moduleColors, stringsAsFactors = FALSE)
#write.csv(x = tx2module, file = "Cohort77_Concensus_tx2module.csv", quote = FALSE)

```

## Relate ME to external traits for each individual set
```{r, echo=FALSE, message=FALSE, warning = FALSE, fig.height=8, fig.width=16}

exprSize = checkSets(multiExpr)
nSets = exprSize$nSets;

# Set up variables to contain the module-trait correlations
moduleTraitCor = list();
moduleTraitPvalue = list();


Traits[[1]]$data$Mean_TTR <- as.numeric(Traits[[1]]$data$Mean_TTR)
Traits[[2]]$data$Mean_TTR <- as.numeric(Traits[[2]]$data$Mean_TTR)

# Calculate the correlations
for (set in 1:nSets)
{
moduleTraitCor[[set]] = cor(consMEs[[set]]$data, Traits[[set]]$data, use = "p");
moduleTraitPvalue[[set]] = corPvalueFisher(moduleTraitCor[[set]], exprSize$nSamples[set]);
}

# Convert numerical lables to colors for labeling of modules in the plot
MEColors = labels2colors(as.numeric(substring(names(consMEs[[1]]$data), 3)));
MEColorNames = paste("ME", MEColors, sep="");
# Open a suitably sized window (the user should change the window size if necessary)

pdf(file = "../figures/ModuleTraitRelationships_by_tissue.pdf", wi = 16, he = 7);
# Plot the module-trait relationship table for set number 1

par(mfrow = c(1,2))

set = 1
textMatrix = paste(signif(moduleTraitCor[[set]], 2), "\n(",
signif(moduleTraitPvalue[[set]], 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor[[set]])
par(mar = c(6, 8.8, 3, 2.2));
labeledHeatmap(Matrix = moduleTraitCor[[set]],
xLabels = names(Traits[[set]]$data),
yLabels = MEColorNames,
ySymbols = MEColorNames,
colorLabels = FALSE,
colors = blueWhiteRed(50),
textMatrix = textMatrix,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),
main = paste("Module--trait relationships in", setLabels[set]))

# Plot the module-trait relationship table for set number 2
set = 2
textMatrix = paste(signif(moduleTraitCor[[set]], 2), "\n(",
signif(moduleTraitPvalue[[set]], 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor[[set]])

#pdf(file = "Plots/ModuleTraitRelationships-male.pdf", wi = 10, he = 7);
par(mar = c(6, 8.8, 3, 2.2));

labeledHeatmap(Matrix = moduleTraitCor[[set]],
xLabels = names(Traits[[set]]$data),
yLabels = MEColorNames,
ySymbols = MEColorNames,
colorLabels = FALSE,
colors = blueWhiteRed(50),
textMatrix = textMatrix,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),

main = paste("Module--trait relationships in", setLabels[set]))

dev.off()

par(mfrow = c(1,1))

```

## Relate ME to traits in concensus set
```{r, echo=FALSE, message=FALSE, warning = FALSE, fig.width=3, fig.height=4}
# Initialize matrices to hold the consensus correlation and p-value
consensusCor = matrix(NA, nrow(moduleTraitCor[[1]]), ncol(moduleTraitCor[[1]]));
consensusPvalue = matrix(NA, nrow(moduleTraitCor[[1]]), ncol(moduleTraitCor[[1]]));
# Find consensus negative correlations
negative = moduleTraitCor[[1]] < 0 & moduleTraitCor[[2]] < 0;
consensusCor[negative] = pmax(moduleTraitCor[[1]][negative], moduleTraitCor[[2]][negative]);
consensusPvalue[negative] = pmax(moduleTraitPvalue[[1]][negative], moduleTraitPvalue[[2]][negative]);
# Find consensus positive correlations
positive = moduleTraitCor[[1]] > 0 & moduleTraitCor[[2]] > 0;
consensusCor[positive] = pmin(moduleTraitCor[[1]][positive], moduleTraitCor[[2]][positive]);
consensusPvalue[positive] = pmax(moduleTraitPvalue[[1]][positive], moduleTraitPvalue[[2]][positive]);

textMatrix = paste(signif(consensusCor, 2), "\n(",
signif(consensusPvalue, 1), ")", sep = "");
textMatrix <-textMatrix %>% str_replace_all(string =., pattern = "NA\\n\\(NA\\)", replacement = "No Concensus")
dim(textMatrix) = dim(moduleTraitCor[[set]])


pdf(file = "../figures/ModuleTraitRelationships_consensus.pdf", wi = 6, he = 6);
par(mar = c(2, 6, 3, 2.2));
labeledHeatmap(Matrix = consensusCor, 
               xLabels = names(Traits[[set]]$data),
               yLabels = MEColorNames,
               ySymbols = stringr::str_remove(MEColorNames, "ME"),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               cex.lab = 0.75,
               xLabelsAngle = 0,
               xLabelsAdj = 0.5,
               zlim = c(-1,1),
               main = paste("Consensus module--trait relationships across\n",
                            paste(setLabels, collapse = " and "))
               )
dev.off()

pdf(file = "../figures/ModuleTraitRelationships_consensus_Age_NC.pdf", wi = 3, he = 4);
par(mar = c(2, 6, 1, 1));
labeledHeatmap(Matrix = consensusCor,
               xLabels = names(Traits[[set]]$data),
               yLabels = MEColorNames,
               ySymbols = stringr::str_remove(MEColorNames, "ME"),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.6,
               cex.lab = 0.75,
               xLabelsAngle = 0,
               xLabelsAdj = 0.5,
               zlim = c(-1,1),
               showCols = 4,
               showRows = 1:12,
               main = NULL
               )
dev.off()



```


#Get hubs
```{r get hubs}


hubs0 <- WGCNA::chooseTopHubInEachModule(datExpr = t(exprDat), colorh = moduleColors, power = 12, type = "signed")

hubs <- data.frame(module = names(hubs0), tx = hubs0, stringsAsFactors = FALSE) %>%
  left_join(., AcTxAnot %>% select(tx, Gene_names, Entry, Protein_names)) %>%
  unique() %>%
  #inner_join(., exprDat %>% as.data.frame() %>% rownames_to_column(var = "tx")) %>%
  inner_join(.,
             data.frame(ageCor = consensusCor[,4], 
                        pval = round(consensusPvalue[,4], digits = 3), 
                        module = MEColorNames %>% stringr::str_replace(string = ., pattern = "ME(.*)","\\1"),
                        stringsAsFactors = FALSE)
  )



write.csv(x = hubs, file = "../results/concensus_hubs.csv", quote = FALSE)
  

scaled_hubs <- hubs
scaled_hubs[,6:(length(scaled_hubs)-2)] <- t(scale(t(scaled_hubs[,6:(length(scaled_hubs)-2)])))
scaled_hubs_long <- scaled_hubs %>% 
  gather(key = "Seq_ID", value = "expr", -c(tx, module, ageCor, pval, Protein_names, Entry, Gene_names)) %>%
  inner_join(samples%>% dplyr::select(Seq_ID, Age))

scaled_hubs %>% mutate(., datText = paste0(Gene_names, "\n", ageCor, "\n", pval))

scaled_hubs %>% dplyr::select(module, tx, Gene_names, ageCor, pval) %>% filter(pval < 0.05)

scaled_hubs_long %>%
  ggplot(data =.,  aes(x = Age, y = expr, color = module)) + 
  geom_smooth(method = "loess") +
  facet_wrap(facets = "module") +
  scale_color_manual(values = MEColorNames %>% stringr::str_remove(.,pattern = "ME")) +
  theme(legend.position = "none")




```


#Make Cytoscape files per module
```{r, eval = FALSE}

lnames = load(file = "../data/r_data/Consensus_TOM.R")

exportAll2Cytoscape(cols = moduleColors, TOM = consensusTOM, data =multiExpr[[1]]$data, minweight = 0.9)


rm(consensusTOM)
```

#Make DyNetViewer files per module
```{r, eval = FALSE}

exprDat = t(rbind(multiExpr[[1]]$data,multiExpr[[2]]$data))
exprDat <- as.data.frame(exprDat)

genCytoscapeTPM(abundances = exprDat, 
                group = factor(kron_samples$Age, levels = c("6","7","8","9","10","11","12")), 
                cols = moduleColors)






```

#Tree plot
```{r, echo=FALSE, message=FALSE, warning = FALSE}
pdf(file = "../figures/ModuleTraitRelationships_tree_plot.pdf", wi = 6, he = 5);
plotDendroAndColors(consTree, moduleColors,
"Merged",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
dev.off()
```


#Eigengene relationships
```{r, echo=FALSE, message=FALSE, warning = FALSE, fig.height = 9, fig.width = 7}

# Recalculate module eigengenes
#MEs = rbind(consMEs[[1]]$data,consMEs[[2]]$data)
MEs = moduleEigengenes(rbind(multiExpr[[1]]$data, multiExpr[[2]]$data),
                       moduleColors)$eigengenes
# Isolate weight from the clinical traits
age = as.data.frame(c(Traits[[1]]$data$Age,Traits[[2]]$data$Age))
names(age) = "Age"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, age))
# Plot the relationships among the eigengenes and the trait
sizeGrWindow(5,7.5);
par(cex = 0.9)

pdf(file = "../figures/Cohort77_Consensus_EigenGene2Age_Dendro.pdf", width = 7, height = 7)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,4.2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle
= 90, setMargins = TRUE)
dev.off()


###Individual correlations for both in set
colnames(consMEs[[1]]$data) <- MEColorNames
colnames(consMEs[[1]]$averageExpr) <- MEColorNames
colnames(consMEs[[1]]$varExplained) <- MEColorNames
colnames(consMEs[[2]]$data) <- MEColorNames
colnames(consMEs[[2]]$averageExpr) <- MEColorNames
colnames(consMEs[[2]]$varExplained) <- MEColorNames

#pdf(file = "../figures/Cohort77_Consensus_Full_Meta.pdf")
plotEigengeneNetworks(multiME = consMEs, 
                      printAdjacency = TRUE,
                      marDendro = c(0,4,1,2),
                      cex.adjacency = 0.25,
                      barplotErrors = TRUE,
                      Letters = TRUE,
                      setLabels = setLabels, 
                      excludeGrey = TRUE, 
                      plotHeatmaps = TRUE, 
                      signed = TRUE, 
                      colorLabels = TRUE,
                      coloredBarplot = TRUE,
                      )
#dev.off()

eigenCor <- cor(MEs) %>% 
  as.data.frame() %>%
  gather(key= "module", value = "correlation") %>%
  dplyr::filter(correlation != 1) %>%
  mutate(magnitude = abs(correlation))


moduleColor2Label <- cbind(moduleColors, moduleLabels) %>% unique()


```

# Gene relationship to trait and important modules: Gene Significance and ModuleMembership
```{r, echo=FALSE, message=FALSE, warning = FALSE}
# Define variable weight containing the weight column of datTrait
age = as.data.frame(c(Traits[[1]]$data$Age,Traits[[2]]$data$Age))
names(age) = "Age"
# names (colors) of the modules
modNames = substring(names(MEs), 3)
geneModuleMembership = as.data.frame(cor(t(exprDat), MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), 73));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(t(exprDat), age, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), 73));
names(geneTraitSignificance) = paste("GS.", names(age), sep="");
names(GSPvalue) = paste("p.GS.", names(age), sep="");

write.csv(x =as.data.frame(cbind(geneTraitSignificance, GSPvalue)), file = "../results/consensus_GeneTrait_Significance.csv", quote = FALSE)

lapply(modNames, function(x){
  moduleGenes = moduleColors==x
  ##module specific genes
  mm <- data.frame(
    membership = geneModuleMembership[moduleGenes, paste0("MM",x)],
    MM_pvalue = MMPvalue[moduleGenes, paste0("p.MM",x)],
    TAS = geneTraitSignificance[moduleGenes, "GS.Age"],
    TAS_pvalue = GSPvalue[moduleGenes, "p.GS.Age"]
  )
  rownames(mm) <- rownames(geneModuleMembership[moduleGenes,])
  mm <- mm %>% rownames_to_column(var = "tx")
  mm <- mm %>%
    left_join(., AcTxAnot %>% dplyr::select(tx, Entry, Gene_names, Protein_names)) %>%
    unique()
  ##top module membership genes
  mmtt <- mm %>%
    dplyr::filter( abs(membership) > quantile(x = abs(mm$membership), probs = 0.9))
  ##membership of all genes
  mm_all <- data.frame(
    membership = geneModuleMembership[, paste0("MM",x)],
    pvalue = MMPvalue[, paste0("p.MM",x)]
  )
  rownames(mm_all) <- rownames(geneModuleMembership[,])
  mm_all <- mm_all %>% rownames_to_column(var = "tx")

  #print(c(nrow(mm), nrow(mmtt)))
  write.csv(x = mm, file = paste0("../results/MM/consensus_moduleMembership_",x,".csv"))
  write.csv(x = mmtt, file = paste0("../results/MM/consensus_moduleMembership_",x,"_topTenPct.csv"))
  write.csv(x = mm_all, file = paste0("../results/MM/consensus_moduleMembership_",x,"_all.csv"))
 })




```

#visualize GS and MM
```{r GS by MM plots for Age, echo=FALSE, message=FALSE, warning = FALSE, fig.width= = 7}
geneTraitSignificance.age <- lapply(X = unique(modNames), FUN = function(x){
column = match(x, modNames)
moduleGenes = moduleColors==x
trait = as.data.frame(c(Traits[[1]]$data[,"Age"],Traits[[2]]$data[,"Age"]))
geneTraitSignificance <- as.data.frame(cor(t(exprDat), trait, use = "p", )) %>%
  rownames_to_column("tx")
geneTraitSignificance <- geneTraitSignificance[moduleGenes,]
geneTraitSignificance[,2] = abs(geneTraitSignificance[,2])
colnames(geneTraitSignificance) = c("tx","GS")
geneTraitSignificance<- geneTraitSignificance %>% mutate(module = x, MM = abs(geneModuleMembership[moduleGenes, column]))
geneTraitSignificance
}) %>% do.call("rbind",.)

geneTraitSignificance.age <- geneTraitSignificance.age %>% filter(module != "grey")

geneTraitSignificance.age$module <- factor(geneTraitSignificance.age$module, levels = c("purple","royalblue","saddlebrown",
                                                "greenyellow","orange","darkgreen","pink",
                                                "steelblue","violet","paleturquoise","blue","green"))

labels <- geneTraitSignificance.age %>% group_by(module) %>%
  summarise(cor = round(cor(GS, MM, use = "p"), 2), 
            count = n(),
            p = formatC(corPvalueStudent(cor, sum(is.finite(MM) & is.finite(GS))), format = "e", digits = 1)
            ) %>%
  mutate(label = paste0(module, " , n= ", count, "\n", "cor = ", cor, ", p = ", p))

module_labeller <- function(variable,value){
  return((labels %>% filter(module == value))$label)
}



geneTraitSignificance.age %>%
  ggplot(data = ., aes( x = MM, y = GS, color = module, fill = module)) +
  geom_point(alpha = 1) +
  geom_smooth(method = "lm", color = "black", se = FALSE, show.legend = FALSE, size = 1.25) +
  facet_wrap(facets = "module", labeller=module_labeller) +
  scale_color_manual(values = levels(labels$module)) +
  scale_x_continuous(breaks = seq(0,0.9,.2))+
  labs(x = "Module Membership", y = "Transcript Significance for Age") +
  theme_bw() +
  theme(
    legend.position = "none",
    panel.grid.minor = element_blank(),
    panel.grid.major = element_line(color = "white"),
    axis.text = element_text(size = 10, color = "black", face = "bold"),
    axis.title = element_text(size = 12, color = "black", face = "bold"),
    strip.background = element_rect(colour="black", fill= "grey90", size = 1),
    strip.text.x = element_text(size = 10, color = "black", face = "bold"),
    panel.border = element_rect(size = 1, color = "black"),
    panel.background = element_rect(fill = "grey95")
  ) #+   geom_text(data = labs, mapping = aes(x = x, y = y, label = meta), fontface = "bold", color = "black", inherit.aes = FALSE)

ggsave(file = "../figures/Cohort_77_ModuleMembership_GeneSignificance.pdf", width = 180, units = "mm")




meta <-  c(rep("meta1",3), rep("meta2",4), rep("meta3",5))
names(meta) <- c("purple","royalblue","saddlebrown",
                                                "greenyellow","orange","darkgreen","pink",
                                                "steelblue","violet","paleturquoise","blue","green")
labs <- as.data.frame(meta) %>% rownames_to_column("module") %>% mutate( x = 0.15, y = 0.725)

```


```{r GS to MM plots for all traits, echo=FALSE, message=FALSE, warning = FALSE, eval = FALSE}

geneTraitSignificance.all <- lapply(X = unique(modNames), FUN = function(x){
column = match(x, modNames)
moduleGenes = moduleColors==x
lapply(X = colnames(Traits[[1]]$data), FUN = function(y){
trait = as.data.frame(c(Traits[[1]]$data[,y],Traits[[2]]$data[,y]))
geneTraitSignificance <- as.data.frame(cor(t(exprDat), trait, use = "p", )) %>%
  rownames_to_column("tx")
geneTraitSignificance <- geneTraitSignificance[moduleGenes,]
geneTraitSignificance[,2] = abs(geneTraitSignificance[,2])
colnames(geneTraitSignificance) = c("tx",paste0(y))
geneTraitSignificance<- geneTraitSignificance %>% mutate(module = x, MM = abs(geneModuleMembership[moduleGenes, column]))
geneTraitSignificance
}) %>% purrr::reduce(full_join, by = c("tx", "module", "MM"))
}) %>% do.call("rbind",.) %>%
  pivot_longer(., cols = c(Weight, Mean_TTR, Mean_TWRT, Age), names_to = "Trait", values_to = "GS")

labels <- geneTraitSignificance.all %>% group_by(module, Trait) %>% 
  summarise(cor = cor(GS, MM, use = "p"), 
            count = n(),
            p = corPvalueStudent(cor, sum(is.finite(MM) & is.finite(GS)))
            ) %>%
  mutate(label = paste0(module, "\n", "cor = ", cor, "\np = ", p),
         category = paste0(module, Trait))

geneTraitSignificance.all %>%
  ggplot(data = ., aes( x = GS, y = MM, color = module, fill = module)) +
  geom_point() +
  geom_smooth(method = "lm", color = "black", se = FALSE, show.legend = FALSE) +
   geom_text(
  data    = labels,
  mapping = aes(x = -Inf, y = -Inf, label = label),
  hjust   = -0.1,
  vjust   = -1
) +
  facet_grid(module ~ Trait, ) +
  scale_color_manual(values = sort(unique(modNames) ) )



```



#Module Eigengene expression trends as function of Age
```{r, echo=FALSE, message=FALSE, warning = FALSE, fig.height=8, fig.width=6}

supp.labs <- as.vector(paste0(MEColorNames, "\n",
                                      "cor = ", signif(consensusCor[,4], 2),"\n", 
                                      "p = ", signif(consensusPvalue[,4], 1)))
names(supp.labs) <- (ME = MEColorNames)

d <- rbind(consMEs[[1]]$averageExpr,consMEs[[2]]$averageExpr) %>% 
  as.data.frame() %>%
  #dplyr::select(-c("MEgrey")) %>%
  rownames_to_column("Seq_ID") %>%
  gather(key = "Module", value = "ME", -c(Seq_ID)) %>%
  mutate(modColors = factor(stringr::str_remove_all(Module,"ME"))) %>%
  inner_join(samples %>% select(Seq_ID, Age)) %>%
  filter(modColors != "grey") 
  ggplot(data =d,  aes(x = Age, y = ME, color = modColors, fill = modColors)) + 
  #geom_rect(inherit.aes = FALSE, mapping = aes(xmin = 9, xmax = 10, ymin = -Inf, ymax = Inf), fill = "grey90")+
  geom_smooth(method = "loess", size = 1.25, alpha = 1, color = "black") +
  geom_vline(xintercept = 9, linetype = 2)+
  geom_vline(xintercept = 10, linetype = 2) +
  theme(legend.position = "none") +
  labs(y = "Scaled Expression", x = "Age (months)", title = "Module Eigengene Expression with Age") +
  scale_fill_manual(values = as.vector(sort(unique(d$modColors)))) +
  facet_wrap(~Module, labeller = labeller(Module = supp.labs))

#ggsave(filename = "Cohort77_consensus_EigengeneExpr.png", device = "png", width = 10, height = 10, units = "in", dpi = 300)
#ggsave(filename = "Cohort77_consensus_EigengeneExpr.eps", device = "eps", width = 6, height = 8, units = "in")

  

```

```{r Eigengene Trajectories Plot}


MEs <- rbind(consMEs[[1]]$averageExpr,consMEs[[2]]$averageExpr) %>% 
  as.data.frame() %>%
  #dplyr::select(MEorange, MEpink, MEroyalblue, MEsaddlebrown) %>%
  rownames_to_column(var = "Seq_ID") %>%
  inner_join(., samples %>%
               select(Age, Seq_ID, Tissue))

ME2 <- MEs %>%
  gather(key = "module", value = "Expr", -c("Seq_ID", "Age", "Tissue")) %>%
  mutate( module_color = stringr::str_remove(string = module, pattern = "ME"))

ME2 %>%
  filter(module != "MEgrey") %>%
  group_by(Age, Tissue) %>%
  ggplot(data =., aes(x = as.factor(Age), y = Expr, fill = module)) + 
  geom_boxplot(alpha = 0.5) + facet_wrap(facets = "module") +
  scale_fill_manual(values = sort(unique((ME2 %>% filter(module != "MEgrey"))$module_color))) +
  labs( x = "Age (months)", y = "Scaled Expression") +
  guides(fill = FALSE) +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    axis.text = element_text(size = 10, color = "black", face = "bold"),
    axis.title = element_text(size = 12, color = "black", face = "bold"),
    strip.background = element_rect(colour="black", fill= "grey90", size = 1), strip.placement = "inside",
    strip.text.x = element_text(size = 12, color = "black", face = "bold"),
    panel.border = element_rect(size = 1),
    panel.background = element_rect(fill = "white")
  )
ggsave(filename = "../figures/Cohort_77_Consensus_ModuleEigengene_boxplots.pdf", dpi = 300, width = 180, units = "mm")

ME2$module <- factor(x = ME2$module, levels = c("MEpurple","MEroyalblue","MEsaddlebrown",
                                                "MEgreenyellow","MEorange","MEdarkgreen","MEpink",
                                                "MEsteelblue","MEviolet","MEpaleturquoise","MEblue","MEgreen"))
ME2$module_color <- factor(x = ME2$module_color , levels = c("purple","royalblue","saddlebrown",
                                                "greenyellow","orange","darkgreen","pink",
                                                "steelblue","violet","paleturquoise","blue","green"))

ME2 %>% 
  filter(module != "MEgrey") %>%
  group_by(Age, Tissue) %>%
  ggplot(data =., aes(x = Age, y = Expr, fill = module)) + 
  geom_hline( yintercept = 0, size = 0.6) +
  geom_smooth( color = "black", alpha = .5) + 
  geom_vline( xintercept = 9, linetype = 2) +
  geom_vline( xintercept = 10, linetype = 2) +
  guides(fill = FALSE) +
  facet_wrap(facets = "module") +
  #scale_fill_manual(values = sort(unique((ME2 %>% filter(module != "MEgrey"))$module_color))) +
  scale_fill_manual(values = levels(ME2$module_color)) +
  labs( x = "Age (months)", y = "Scaled Expression") +
  scale_x_continuous(breaks = seq(6,12,1)) +
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_line(color = "white"),
    axis.text = element_text(size = 10, color = "black", face = "bold"),
    axis.title = element_text(size = 12, color = "black", face = "bold"),
    strip.background = element_rect(colour="black", fill= "grey90", size = 1),
    strip.text.x = element_text(size = 12, color = "black", face = "bold"),
    panel.border = element_rect(size = 1, color = "black"),
    panel.background = element_rect(fill = "grey95")
  ) +
  geom_text(data = labs, aes(x = x, y = y, label = meta), fontface = "bold", color = "black")
  
ggsave(filename = "../figures/Cohort_77_Consensus_ModuleEigengene_Trajectores.pdf", dpi = 300, width = 180, units = "mm")
  
  
meta <-  c(rep("meta1",3), rep("meta2",4), rep("meta3",5))
names(meta) <- c("MEpurple","MEroyalblue","MEsaddlebrown",
                                                "MEgreenyellow","MEorange","MEdarkgreen","MEpink",
                                                "MEsteelblue","MEviolet","MEpaleturquoise","MEblue","MEgreen")
labs <- as.data.frame(meta) %>% rownames_to_column("module") %>% mutate( x = 11.1, y = -1)


```


#Plot Eigengenes expression per sample
```{r, echo=FALSE,warning=FALSE,message=FALSE, fig.width=7}

MEs <- rbind(consMEs[[1]]$averageExpr,consMEs[[2]]$averageExpr) %>% 
  as.data.frame()

# module = "cyan"
# data.frame(Seq_ID = rownames(MEs), ME_Expr = MEs[,paste0("ME", module)]) %>% 
#   select(paste0("ME", module)) %>%
#   rownames_to_column(var = "Seq_ID") %>%
#   inner_join(., kron_samples %>% select(Age, Seq_ID)) %>%
#     ggplot(data =., aes(x = Seq_ID, y = ME_Expr)) + geom_bar(stat = "identity")
# 
# 
module = "orange"
dats <- data.frame(Seq_ID = rownames(MEs), ME_Expr = MEs[,paste0("ME", module)]) %>%
  inner_join(., samples%>% select(Age, Seq_ID, Tissue))
dats$Seq_ID <- factor(x = dats$Seq_ID, levels = dats$Seq_ID[order(dats$Age)])
# 
# 
# ggplot(data =dats, aes(x = Seq_ID, y = ME_Expr)) + geom_bar(stat = "identity", fill = module, color = "black") + facet_wrap(facets = "Tissue")
# 
# dats$Seq_ID <- factor(x = dats$Seq_ID, levels = dats$Seq_ID[order(dats$Tissue)])
# ggplot(data =dats, aes(x = Seq_ID, y = ME_Expr)) + geom_bar(stat = "identity", fill = module, color = "black")
# 
# 
# ggplot(data =dats, aes(x = as.factor(Age), y = ME_Expr)) + geom_boxplot(fill = module, color = "black") + facet_wrap(facets = "Tissue")

### Individual barplots for tissue and module

library(ggpubr)

bsc_annot <- data.frame(
xmin= c(0,12,23)+0.5,
xmax=c(6,17,29)+0.5,
ymin = rep(-Inf,3),
ymax = rep(Inf,3)
)
bsc_text <- data.frame(text_x = c(3.5,9.5,15,20.5,26.5,32), text_y = rep(-1.5,6), text =c(7:12))


pvc_annot <- data.frame(
xmin= c(0,11,21,33)+0.5,
xmax=c(5,16,27,39)+0.5,
ymin = rep(-Inf,4),
ymax = rep(Inf,4)
)
pvc_text <- data.frame(text_x = c(2.5,8.5,13.5,18.5,24,30,36)+0.5, text_y = rep(-1.5,7), text =c(6:12))

unique(moduleColors)

p4 <- lapply(X = c("pink", "orange", "royalblue", "saddlebrown"), FUN = function(module){
#p4 <- lapply(X = c(unique(MEColors[MEColors !="grey"])), FUN = function(module){
  
  dats <- data.frame(Seq_ID = rownames(MEs), ME_Expr = MEs[,paste0("ME", module)]) %>% 
    inner_join(., samples %>% select(Age, Seq_ID, Tissue))
  dats$Seq_ID <- factor(x = dats$Seq_ID, levels = dats$Seq_ID[order(dats$Age)])
  
  dats_bsc = dats %>% filter(Tissue == "BSC") %>% droplevels()
  dats_pvc = dats %>% filter(Tissue == "PVC") %>% droplevels()
  
  p1 <- ggplot(data = dats_bsc, aes(x = as.numeric(Seq_ID), y = ME_Expr))+ 
    geom_rect(data = bsc_annot, inherit.aes=FALSE,
              aes(xmin=xmin,xmax=xmax,ymin=ymin,ymax=ymax),fill = "grey75", alpha=1) + 
    geom_bar(stat = "identity", fill = module, color = "black") +
    geom_text(data = bsc_text, inherit.aes=FALSE, aes(x = text_x, y = text_y, label = text), nudge_y = 0.1) +
    scale_y_continuous(limits = c(-1.5,1.5)) +
    geom_hline(yintercept = 0) +
    theme_classic2() +
    theme(axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),
          axis.text.x=element_blank(),
          axis.title = element_blank()
    ) +
    labs(title = "BSC")
  
  p2 <- ggplot(data = dats_pvc, aes(x = as.numeric(Seq_ID), y = ME_Expr))+ 
    geom_rect(data = pvc_annot, inherit.aes=FALSE,
              aes(xmin=xmin,xmax=xmax,ymin=ymin,ymax=ymax),fill = "grey75", alpha=1) + 
    geom_bar(stat = "identity", fill = module, color = "black") +
    geom_text(data = pvc_text, inherit.aes=FALSE, aes(x = text_x, y = text_y, label = text), nudge_y = 0.1) +
    scale_y_continuous(limits = c(-1.5,1.5)) +
    geom_hline(yintercept = 0) +
    theme_classic2() +
    theme(axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),
          axis.text.x=element_blank(),
          axis.title = element_blank()
    ) +
    labs(title = "PVC")
  
  
  p3 <- ggpubr::ggarrange(plotlist = list(p1,p2), nrow = 1)
  annotate_figure(p3, 
                  top = text_grob(paste0("ME",module), color = "black", face = "bold", size = 16),
  )
}
)

p5 <- ggpubr::ggarrange(plotlist = p4) %>%
annotate_figure(., 
                  #top = text_grob("Age Associated Module Eigengene Expression", color = "black", face = "bold", size = 16),
                  bottom = text_grob("Age (Months)", color = "black", face = "bold", size = 12),
                  left = text_grob("Scaled Expression", color = "black", face = "bold", size = 12,rot = 90))

p5


```



#Heatmap of Aging clusters
### set up heatmap data
```{r, echo=FALSE,warning=FALSE,message=FALSE, fig.width=12,fig.height=8}


MEs <- rbind(consMEs[[1]]$averageExpr,consMEs[[2]]$averageExpr) %>% 
  as.data.frame() %>%
  dplyr::select(MEorange, MEpink, MEroyalblue, MEsaddlebrown) %>%
  rownames_to_column(var = "Seq_ID") %>%
  inner_join(., samples %>%
               select(Age, Seq_ID, Tissue))

m1 = MEs %>% filter(Tissue == "BSC") %>% select(-c("Age","Seq_ID","Tissue"))
m2 = MEs %>% filter(Tissue == "PVC") %>% select(-c("Age","Seq_ID","Tissue"))

modcols <- unlist(stringr::str_remove(string = colnames(select(MEs, -c(Age, Seq_ID, Tissue))), pattern = "ME"))
names(modcols) <- colnames(select(MEs, -c(Age, Seq_ID, Tissue)))

col_fun = colorRamp2(c(6, 7, 8, 9, 10, 11, 12), c("tomato","darkorange","gold","green3","turquoise","cornflowerblue","purple"))
col_fun2 = colorRamp2(c(1,2), c("cadetblue","indianred1"))
col_fun3 = colorRamp2(c(-1.5,0,1.5), c("purple","black","yellow"))





```


### Try another heatmap method
```{r, echo=FALSE,warning=FALSE,message=FALSE, fig.width=12,fig.height=8}
ha1 <- rowAnnotation(Age = anno_simple(MEs$Age, col = col_fun),
                     show_annotation_name = FALSE)
ha2 <- rowAnnotation(Age = anno_simple(as.numeric(as.factor(MEs$Tissue)), col = col_fun2),
                     show_annotation_name = FALSE)
ha3 <- columnAnnotation(
  Names = anno_text(x = modcols, rot = 0, just = "center", location = 0.5, gp = gpar(fontface = "bold")),
  MEs = anno_simple(
    colnames(select(MEs, -c(Age, Seq_ID, Tissue))), 
    col = modcols,
     gp = gpar()),
  bsc_barplot = anno_empty(),
  show_annotation_name = FALSE,
  height = unit(4, "cm"))


h1 <- Heatmap(matrix = m1,
              name = "BSC",
              cluster_rows = FALSE,
              #row_split = MEs$Age,
              row_split = (MEs %>% filter(Tissue == "BSC"))$Age,
              row_title_rot = 0,
              column_title_gp = gpar(fontface = "bold"),
              top_annotation = ha3,
              #left_annotation = rowAnnotation(Age = anno_simple( (MEs %>% filter(Tissue == "BSC"))$Tissue, 
              #                                                  col =setNames("cadetblue", "BSC")),
              #       show_annotation_name = FALSE),
              #right_annotation = rowAnnotation(BSC = anno_empty(border = FALSE)),
              show_column_names = FALSE,
              column_split = c(rep("Aging",2),rep("Anti-Aging",2)),
             # row_title = "BSC",
              show_heatmap_legend = FALSE,
              col = col_fun3
)
h2 <- Heatmap(matrix =m2,
              name = "PVC",
              cluster_rows = FALSE,
              #row_split = MEs$Age,
              row_split = (MEs %>% filter(Tissue == "PVC"))$Age,
              row_title_rot = 0,
              column_title_gp = gpar(fontface = "bold"),
              show_column_names = FALSE,
              #row_title = "PVC",
              top_annotation = columnAnnotation(pvc_barplot = anno_empty(), show_annotation_name = FALSE, height = unit(2.25, "cm")),
              #right_annotation = rowAnnotation(PVC = anno_empty(border = FALSE)),
              show_heatmap_legend = FALSE,
              col = col_fun3
)
lgd = Legend(col_fun = col_fun3, 
             title = "Scaled Expression", at = c(-1.5, -0.75, 0, 0.75, 1.5), 
             direction = "horizontal", 
             title_position = "topcenter"
)

```

#Make Eigengene Heatmap 
```{r, echo=FALSE,warning=FALSE,message=FALSE, fig.width=12,fig.height=8, fig.height=8, fig.width=6}
# draw(h1 %v% h2 , annotation_legend_list = lgd, annotation_legend_side = "bottom")
# list_components()

#lapply(row_order(h1), FUN = function(x){length(x)}) %>% unlist
draw(h1 %v% h2 , annotation_legend_list = lgd, annotation_legend_side = "bottom", padding = unit(c(2, 8, 2, 2), "mm"))
#aging bsc annotations
decorate_annotation("bsc_barplot", slice = 1, {
  pushViewport(viewport(xscale = c(0.5, 2.5),yscale = c(-2,2)))
  for(i in seq_along(1:2)) {
    grid.rect(x = ((c(6,17,28)-0.25)/34)-0.4+2-i+1, width = c(6,6,6)/34, default.units = "native", gp=gpar(fill = "grey90", col = NA))
    grid.rect(x = 2-i+1, width = 1, default.units = "native", gp= gpar(fill = NA))
  }
  for(i in seq_along(1:2)){
    grid.text(x = (c(3,9,14.5,20,25.5,31))/(34*2)-1/2+i/2, y =(rep(0.95,6)), label = c("7","8","9","10","11","12"), gp = gpar(cex=0.5))
  }
  grid.lines( x =c(0, 1), y=c(0.25,0.25), gp = gpar(lty = 2, lwd = 1))
  grid.lines( x =c(0, 1), y=c(0.75,0.75), gp = gpar(lty = 2, lwd = 1))
  
  #royal blue
  grid.rect(x = (((1:nrow(m1))-0.5)/nrow(m1))-0.5+3-3+1, width =1/nrow(m1), height = m1[ ,"MEorange"], just = "bottom",
            gp = gpar(fill = "orange"), default.units = "native")
  #pink
  grid.rect(x = (((1:nrow(m1))-0.5)/nrow(m1))-0.5+3-2+1, width =1/nrow(m1), height = m1[ ,"MEpink"], just = "bottom",
            gp = gpar(fill = "pink"), default.units = "native")

})
#antiaging bsc annotations
decorate_annotation("bsc_barplot", slice = 2, {
  for(i in seq_along(1:2)) {
    grid.rect(x = ((c(6,17,28)-0.25)/34)-0.4+2-i+1, width = c(6,6,6)/34, default.units = "native", gp=gpar(fill = "grey90", col = NA))
    grid.rect(x = 2-i+1, width = 1, default.units = "native", gp= gpar(fill = NA))
  }
  
    for(i in seq_along(1:2)){
    grid.text(x = (c(3,9,14.5,20,25.5,31))/(34*2)-1/2+i/2, y =(rep(0.95,6)), label = c("7","8","9","10","11","12"), gp = gpar(cex=0.5))
    }
  
  grid.lines( x =c(0, 1), y=c(0.25,0.25), gp = gpar(lty = 2, lwd = 1))
  grid.lines( x =c(0, 1), y=c(0.75,0.75), gp = gpar(lty = 2, lwd = 1))
  #MEroyalbluee
  grid.rect(x = (((1:nrow(m1))-0.5)/nrow(m1))-0.5+3-3+1, width =1/nrow(m1), height = m1[ ,"MEroyalblue"]/4, just = "bottom",
            gp = gpar(fill = "royalblue"), default.units = "native")
  #MEsaddelbrown
  grid.rect(x = (((1:nrow(m1))-0.5)/nrow(m1))-0.5+3-2+1, width =1/nrow(m1), height = m1[ ,"MEsaddlebrown"]/4, just = "bottom",
            gp = gpar(fill = "saddlebrown"), default.units = "native")

})
#aging pvc annotations
decorate_annotation("pvc_barplot", slice = 1, {
  pushViewport(viewport(xscale = c(0.5, 2.5),yscale = c(-2,2)))
  for(i in seq_along(1:2)) {
    grid.rect(x = ((c(0,11,21,33)-0.25)/39)-0.42+2-i+1, width = c(5,5,6,6)/39, default.units = "native", gp=gpar(fill = "grey90", col = NA))
    grid.rect(x = 2-i+1, width = 1, default.units = "native", gp= gpar(fill = NA))}
  
  for(i in seq_along(1:2)){
    grid.text( x = (c(2.5,8,13.5,18.5,24,30,36))/(39*2)-1/2+i/2, y =(rep(0.925,7)), label = c("6","7","8","9","10","11","12"), gp = gpar(cex=0.5))
  }
  grid.lines( x =c(0, 1), y=c(0.25,0.25), gp = gpar(lty = 2, lwd = 1))
  grid.lines( x =c(0, 1), y=c(0.75,0.75), gp = gpar(lty = 2, lwd = 1))
  #orange
  grid.rect(x = (((1:nrow(m2))-0.5)/nrow(m2))-0.5+3-3+1, width =1/nrow(m2), height = m2[ ,"MEorange"], just = "bottom",
            gp = gpar(fill = "orange"), default.units = "native")
  #pink
  grid.rect(x = (((1:nrow(m2))-0.5)/nrow(m2))-0.5+3-2+1, width =1/nrow(m2), height = m2[ ,"MEpink"], just = "bottom",
            gp = gpar(fill = "pink"), default.units = "native")

})
#antiaging pvc annotations
decorate_annotation("pvc_barplot", slice = 2, {
  pushViewport(viewport(xscale = c(0.5, 2.5),yscale = c(-2,2)))
  for(i in seq_along(1:2)) {
    grid.rect(x = ((c(0,11,21,33)-0.25)/39)-0.42+2-i+1, width = c(5,5,6,6)/39, default.units = "native", gp=gpar(fill = "grey90", col = NA))
    grid.rect(x = 2-i+1, width = 1, default.units = "native", gp= gpar(fill = NA))}
  
  for(i in seq_along(1:2)){
    grid.text( x = (c(2.5,8,13.5,18.5,24,30,36))/(39*2)-1/2+i/2, y =(rep(0.925,7)), label = c("6","7","8","9","10","11","12"), gp = gpar(cex=0.5))
  }
  grid.lines( x =c(0, 1), y=c(0.25,0.25), gp = gpar(lty = 2, lwd = 1))
  grid.lines( x =c(0, 1), y=c(0.75,0.75), gp = gpar(lty = 2, lwd = 1))
  #royalbue
  grid.rect(x = (((1:nrow(m2))-0.5)/nrow(m2))-0.5+3-3+1, width =1/nrow(m2), height = m2[ ,"MEroyalblue"], just = "bottom",
            gp = gpar(fill = "royalblue"), default.units = "native")
  #saddlebrown
  grid.rect(x = (((1:nrow(m2))-0.5)/nrow(m2))-0.5+3-2+1, width =1/nrow(m2), height = m2[ ,"MEsaddlebrown"], just = "bottom",
            gp = gpar(fill = "saddlebrown"), default.units = "native")

})
#rowtitles
decorate_row_title("BSC",
                   slice = 1,
                   {
                     grid.text(label = "BSC", rot = 90, x = -0.5, y = -0.75, gp = gpar(fontface = "bold"))
                     grid.lines( x =c(0, 0), y=c(4, -5.2), gp = gpar(lty = 1, lwd = 2))
                   })
decorate_row_title("PVC",
                   slice = 1,
                   {
                     grid.text(label = "PVC", rot = 90, x = -0.5, y = -1.5, gp = gpar(fontface = "bold"))
                     grid.lines( x =c(0, 0), y=c(4.5, -7.5), gp = gpar(lty = 1, lwd = 2))
                   })


```




#Kegg Enrichment
```{r, fig.width=7}
library(clusterProfiler)

tx2module2KEGG <- tx2module %>%  inner_join(Tx2Prot2Kegg)

lapply( unique(tx2module$module), function(x){
  
  mk <- clusterProfiler::enrichKEGG(
    gene = (tx2module2KEGG %>% filter(ko != "") %>% filter(module == x))$ko,
    organism = "ko",
    pvalueCutoff = 0.05
    
  )
  
  if(nrow(mk[]) > 0){
  emapplot(mk)
  ggsave(file = paste0("../figures/emap_plots/",x, "_enrichKEGG_emapplot.pdf"),device = "pdf", width = 180, units = "mm")
  }
  
  write.csv(x = as.data.frame(mk), file = paste0("../results/cpKEGG/consensus_KEGG_",x,".csv"))
  
})



#MMPvalue %>% rownames_to_column("tx") %>% dplyr::select( tx,paste0("p.MM",x)) %>% arrange_at(.vars = paste0("p.MM",x)) %>%

lapply( unique(tx2module$module), function(x){
  
  mkk <- clusterProfiler::enrichMKEGG(
    gene = (tx2module2KEGG %>% filter(ko != "") %>% filter(module == x))$ko,
    organism = "ko",
    pvalueCutoff = 0.05
    
  )
  
  if(nrow(mkk[]) > 0){
  emapplot(mkk)
  ggsave(file = paste0("../figures/emap_plots/",x, "_enrichMKEGG_emapplot.pdf"),device = "pdf", width = 180, units = "mm")
  }
  
  write.csv(x = as.data.frame(mkk), file = paste0("../results/cpMKEGG/consensus_KEGG_",x,".csv"))
  
})



mk <- clusterProfiler::enrichKEGG(
    gene = (tx2module2KEGG %>% filter(ko != "") %>% filter(module == "royalblue"))$ko,
    organism = "ko",
    pvalueCutoff = 0.05
    
  )
  

emapplot(mk, layout = "nicely")
ggsave("../figures/emap_plots/royalblue_enrichMKEGG_emapplot.pdf")

```

##GO enrichment
```{r cluster profiler GO}
library(org.Hs.eg.db)
tx2module2Uniprot <- inner_join(tx2module, AcTxAnot %>% dplyr::select(., tx, Entry))

lapply(unique(tx2module$module), function(x) {
  gg <- lapply(X = c("BP", "MF", "CC"), function(y) {
    mg <- clusterProfiler::enrichGO(
      gene = (
        tx2module2Uniprot %>% filter(Entry != "") %>% filter(module == x)
      )$Entry,
      ont = y,
      OrgDb = org.Hs.eg.db,
      keyType = "UNIPROT",
      pvalueCutoff = 0.05
    )
    
    mg %>%
      as.data.frame() %>% mutate(ontology = y)
  }) %>% do.call("rbind",.)
  write.csv(x = as.data.frame(gg),
            file = paste0("../results/topGO/consensus_GO_", x, ".csv"))
})

```

```{r GO enrichment function, echo=FALSE, warning=FALSE,message=FALSE}

getGO <- function(uni,
           clusters,
           clust_ID,
           ontology = "BP",
           save = FALSE,
           simple = FALSE) {
    ##make sure user didn't specify an impossible option
    if (!ontology %in% c("BP", "MF", "CC")) {
      print("Unkown ontology selected, using BP")
      ontology = "BP"
    }
    
  genes <- clusters %>% dplyr::filter(module == clust_ID) %>% inner_join(., AcTxAnot) %>% unique()
  genes <- genes$Entry
  
    ## Get GO enrichment
    go <-
      enrichGO(
        gene          = genes,
        universe      = uni,
        OrgDb         = org.Hs.eg.db,
        keyType       = "UNIPROT",
        ont           = ontology,
        pvalueCutoff  = 0.05,
        qvalueCutoff  = 0.05,
        readable      = TRUE
      )
    
    ##Determine if user wants reduced output or not
    if (simple == TRUE) {
      go <-
        go %>% clusterProfiler::gofilter(., 5) %>% as.data.frame() %>% arrange(p.adjust)
      outfile = paste0("../results/cpGO/",
                       clust_ID,
                       "_",
                       ontology,
                       "_GO_level5.txt")
    }
    else{
      go <- go %>% as.data.frame() %>% arrange(p.adjust)
      outfile = paste0("../results/cpGO/",
                       clust_ID,
                       "_",
                       ontology,
                       "_GO.txt")
    }
    if (save == TRUE) {
      write.table(
        x = go,
        file = outfile,
        quote = FALSE,
        row.names = FALSE,
        sep = "\t"
      )
    }
    invisible(go)
  }
    


```


```{r Go enrichment for all clusters, echo=FALSE, warning=FALSE,message=FALSE}

### THIS IS SUPER SLOW, TRY TO DO ONLY ONCE
## //todo OPTIMIZE

library(org.Hs.eg.db)

uni <-  (AcTxAnot %>% inner_join(tx2module)%>% unique())$Entry


lapply(unique(tx2module$module), function(x) {
  print(paste0("Processing:", x))
  enrichment <- lapply(c("BP", "MF", "CC"), function(y) {
    getGO(
      uni = uni,
      clusters = tx2module,
      clust_ID = x,
      ontology = y,
      save = FALSE,
      simple = TRUE
    ) %>%
      as.data.frame() %>% mutate(ontology = y) %>% dplyr::select(ontology, everything())
  }) %>% do.call("rbind", .)
  write.table(
    x = enrichment,
    file = paste0("../results/cpGO/", x, "_allGO.txt"),
    sep = "\t",
    quote = FALSE,
    row.names = FALSE
  )
})




  
```

<!-- Define functions to get maximal log 2 fold change from expression matrix -->
```{r, echo=FALSE, message=FALSE, warning = FALSE}
maxL2FC <- function(x, log2 = FALSE){
  #' Get maximal log2 Fold Change of input vector
  #'
  #' Takes a vector and calculates the maximal log2 fold change of a sequence of values.
  #' If only 2 values are provided, collapses to normal log2 fold change calculation.
  #'
  #' @param x numeric. A vector that is assumed to be in order. Must have at least 2 components.
  #' @param log2 boolean. Specify if input values need to be log transformed, TRUE if yes. Default FALSE.
  #' @return numeric. The maximal log2 fold change of the input vector. 
  
  a = 0
  b = 0
  
  #check user input
  if(!is.numeric(x)){
    stop("Input is not of type numeric")
  }
  if(length(x) < 2){
    stop("Input must have at least 2 values")
  }
  #rising trend
  if( match(max(x), x) > match(min(x), x) ){
    a = min(x)
    b = max(x)
  }
  #falling trend
  else if ( match(min(x), x) > match(max(x), x) ){
    a = max(x)
    b = min(x)
  }
  #log2(value +1) transform to get log fold change
  if(log2 == TRUE){
    log(b+1,2) - log(a+1,2)
  }
  else{
    b-a
  }
  
}


getMaxL2FC <- function(em, byRow = TRUE, log2 = FALSE){
  #' Get maximal log2 Fold Change of input expression matrix
  #'
  #' Takes an expression matrix and calculates the maximal log2 fold change per gene.
  #'
  #' @param me numeric. An expression matrix.
  #' @param byRow boolean. If TRUE, matrixs is gene by sample, ie rows represent genes. FALSE will evaluate by column.
  #' @param log2 boolean. Specify if input values need to be log transformed, TRUE if yes. Default FALSE.
  #' @return numeric. A vector containing the maximal log2 fold change of an input matrix. 
  
  if(! (is.matrix(em) | is.data.frame(em) ) ){
    stop("Input must be an expression matrix.")
  }
  em <- as.matrix(em)
  
  if(byRow == TRUE){
    m = 1L
  }
  else{
    m = 2L
  }
  apply(em, MARGIN = m, FUN = maxL2FC)
  
}


```

```{r}

datBSC <- t(as.matrix(multiExpr[[1]]$data))
datPVC <- t(as.matrix(multiExpr[[2]]$data))

write.csv(x = getMaxL2FC(datBSC), file = "../results/Cohort77_maxL2FC_BSC.csv", quote = FALSE)
write.csv(x = getMaxL2FC(datPVC), file = "../results/Cohort77_maxL2FC_PVC.csv", quote = FALSE)

```




#Sig P and Annot Stats
```{r, echo=FALSE, message=FALSE, warning = FALSE}

load("../data/r_data/metadata.R")
load("../data/r_data/res.bsc.R")
load("../data/r_data/res.pvc.R")

tx2module <- data.frame(tx = rownames(exprDat), module = moduleColors, stringsAsFactors = FALSE)


module_annot <- tx2module %>% left_join(., AcTxAnot %>% select(tx, prot, ko, Entry)) %>%
  inner_join(res.bsc %>% rownames_to_column("tx") %>% select(padj, tx)) %>%
  inner_join(res.pvc %>% rownames_to_column("tx") %>% select(padj, tx), by = "tx",  suffix = c(".bsc", ".pvc"))

module_annot_smry <- module_annot %>% group_by(module) %>%
  summarise(
    n = n(),
    KO = sum(ko != "" & is.na(ko) == FALSE),
    Uniprot = sum(Entry != "" & is.na(Entry) == FALSE),
    BSC = sum(padj.bsc <= 0.01),
    PVC = sum(padj.pvc <= 0.01)
  )

module_annot_smry_prop <- c(module_annot_smry[,3:6] / rep(module_annot_smry[,2],4), module_annot_smry[,1]) %>% 
  as.data.frame()
  


module_annot_smry_prop %>% 
  ggplot(data = ., aes(x = module, y = KO, fill = module)) + 
  geom_bar(stat = "identity", position="dodge", color = "black") +
  scale_fill_manual(values = as.character(module_annot_smry$module)) +
  geom_hline(yintercept = 0.5)+
  labs(x = "Module", y = "Proportion", title = "Proportion of Module Genes KEGG Annotated") +
    theme(axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),
          axis.text.x=element_blank())


module_annot_smry_prop %>% 
  ggplot(data = ., aes(x = module, y = Uniprot, fill = module)) + 
  geom_bar(stat = "identity", position="dodge", color = "black") +
  scale_fill_manual(values = as.character(module_annot_smry$module)) +
  geom_hline(yintercept = 0.5)+
  labs(x = "Module", y = "Proportion", title = "Proportion of Module Genes UNIPROT Annotated") +
    theme(axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),
          axis.text.x=element_blank())

module_annot_smry_prop %>% 
  pivot_longer(data = ., values_to = "Annot", names_to = "val", cols = -c("module")) %>%
  filter(val %in% c('BSC', 'PVC'))%>%
  ggplot(data = ., aes(x = module, y = Annot, fill = module)) + 
  geom_bar(stat = "identity", position="dodge", color = "black") +
  scale_fill_manual(values = as.character(module_annot_smry$module)) +
  geom_hline(yintercept = 0.25) +
  facet_wrap(facets = "val") +
  labs(x = "Module", y = "Proportion", title = "Proportion of Module Genes Significant in Kron 2020") +
    theme(axis.ticks.x=element_blank(),
          axis.line.x=element_blank(),
          axis.text.x=element_blank())

```

# overlap in gene lists of modules and clusters from Kron 2020
```{r, echo=FALSE, message=FALSE, warning = FALSE}

load("../data/r_data/res.pvc.R")
load("../data/r_data/res.bsc.R")

bsc_clust <- c(
  "1" = "B1",
  "2" = "B3",
  "3" = "B2",
  "5" = "B4"
)
res.bsc_clusters$clust <- bsc_clust[as.character(res.bsc_clusters$cluster)]

pvc_clust <- c(
  "1" = "P1",
  "3" = "P2",
  "5" = "P3",
  "4" = "P4",
  "8" = "P5"
)
res.pvc_clusters$clust <- pvc_clust[as.character(res.pvc_clusters$cluster)]

bsc_overalp <- lapply(unique(tx2module$module), FUN = function(x){
  
  lapply(unique(res.bsc_clusters$clust), FUN = function(y){
    
    inner_join( 
      res.bsc_clusters %>% dplyr::filter(clust == y),
      tx2module %>% dplyr::filter(module == x), by = c("genes" = "tx")
      ) %>% nrow() #/ nrow(res.bsc_clusters %>% dplyr::filter(clust == y)) 
    
  }) %>% unlist() %>% round(., 2)
  
}) %>% do.call("rbind",.)

pvc_overalp <- lapply(unique(tx2module$module), FUN = function(x){
  
  lapply(unique(res.pvc_clusters$clust), FUN = function(y){
    
    inner_join( 
      res.pvc_clusters %>% dplyr::filter(clust == y),
      tx2module %>% dplyr::filter(module == x), by = c("genes" = "tx")
      ) %>% nrow() #/ nrow(res.pvc_clusters %>% dplyr::filter(clust == y)) 
    
  }) %>% unlist() %>% round(., 2)
  
}) %>% do.call("rbind",.)

rownames(bsc_overalp) = unique(tx2module$module)
rownames(pvc_overalp) = unique(tx2module$module)

colnames(bsc_overalp) = unique(res.bsc_clusters$clust)
colnames(pvc_overalp) = unique(res.pvc_clusters$clust)

all_overlap <- cbind (bsc_overalp, pvc_overalp)
all_overlap <- all_overlap[, c("B1","B2","B3","B4","P1","P2","P3","P4","P5")]

library(ComplexHeatmap)

Heatmap(all_overlap,
        cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
        grid.text(all_overlap[i, j], x, y) })



inner_join( 
      res.pvc_clusters %>% dplyr::filter(clust == "P4"),
      tx2module %>% dplyr::filter(module == "saddlebrown"), by = c("genes" = "tx")
      )


res.pvc_clusters %>% group_by(clust) %>% summarise(n = n())


```

